src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs

```diff
@@ -71,7 +71,17 @@ private static unsafe void CompletionPortCallback(uint errorCode, uint numBytes,
                 {
                     NetEventSource.Fail(null, $"asyncResult.IsCompleted: {asyncResult}");
                 }
-                if (NetEventSource.IsEnabled) NetEventSource.Info(null, $"errorCode:{errorCode} numBytes:{numBytes} nativeOverlapped:{(IntPtr)nativeOverlapped}");
+                if (NetEventSource.IsEnabled)
+                {
+                    try
+                    {
+                        NetEventSource.Info(null, $"errorCode:{errorCode} numBytes:{numBytes} nativeOverlapped:{(IntPtr)nativeOverlapped}");
+                    }
+                    catch
+                    {
+                        // Do nothing. If NetEventSource fails, we must continue to avoid a deadlock. Completion port handlers must not fail.
+                    }
```

Comment:
If this is an issue, why is it limited to the Info call... why not the Fail call above?  Why not the other use of NetEventSource below?

And what failures is this referring to... just OOM?  It should explicitly handle just OutOfMemoryException then.

This level of reliability, against OOMs encountered when allocating even small objects, is a game of whack-a-mole and unlikely to be successful in the extreme.

I'm also confused by the "deadlock" comments.  If this throws an exception that goes unhandled, won't the process crash?

Reply:
There's a try/catch/ignore on the top of the stack in the callback dispatcher (in native code).

Whether it would have been acceptable to crash the process here is a guessing game to me, but as it stands, it deadlocks the machine it happens on because the callback dispatcher never lets go of the memory. An out of memory later on doesn't have this problem.

In this case I catch anything because any possible failure of NetEventSource would trigger the deadlock.

Reply:
Actually the whack-a-mole is not so bad, because the only place this can happen is inside an IO completion handler or something called by it. Anywhere else appears to just leak the task and the GC eventually cleans it up.

Reply:
> There's a try/catch/ignore on the top of the stack in the callback dispatcher (in native code).

Can you link to it? Thanks.

Reply:
Unfortunately that particular try/catch/ignore is in the Windows source code. It's interesting to note that the portable version of the task scheduler, which is used on all platforms but Windows, works like you say it does (unhandled exception -> crash).

src/vm/comthreadpool.cpp doesn't catch anything in the callback stub and _has another instance of this class of problem_ on line 651.

Reply:
> Unfortunately that particular try/catch/ignore is in the Windows source code.

Can you link to the one in your call stack that's eating your exception? Thanks.

Reply:
I don't understand what you mean. The machine hangs up and when we reboot it we occasionally find OutOfMemoryException in the log file (standard output is redirected to a file) with no stack trace, but most of the time there's nothing there at all. We've been throwing memory at the problem to keep incidents down. But then a sales person runs a memory gobbler (different process) and their personal instance hangs up.

Reply:
> I don't understand what you mean.

I mean you said "There's a try/catch/ignore on the top of the stack in the callback dispatcher (in native code)"... can you please point me to it?

Reply:
On Windows, the managed thread pool uses the native thread pool. The native thread pool swallows exceptions thrown from IO completion ports.

Reply:
> On Windows, the managed thread pool uses the native thread pool. The native thread pool swallows exceptions thrown from IO completion ports.

If I write:
```C#
using System;
using System.Threading;

class Program
{
    static unsafe void Main()
    {
        var o = new Overlapped();
        NativeOverlapped* no = o.Pack(delegate
        {
            throw new OutOfMemoryException();
        }, null);

        ThreadPool.UnsafeQueueNativeOverlapped(no);
        Console.ReadLine();
    }
}
```
the app crashes due to the OutOfMemoryException.

If I change https://github.com/dotnet/coreclr/blob/eb9275aaa7d4e0484f533b1daafb61c70c4dad38/src/System.Private.CoreLib/shared/System/IO/FileStreamCompletionSource.Win32.cs#L129 to just unconditionally throw an OutOfMemoryException, and then run this:
```C#
using System;
using System.IO;
using System.Threading.Tasks;

class Program
{
    static async Task Main()
    {
        string path = Path.GetTempFileName();
        File.WriteAllBytes(path, new byte[10_000_000]);

        var buffer = new byte[100];
        using (var fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 1, useAsync: true))
        {
            int read;
            while ((read = await fs.ReadAsync(buffer)) != 0)
            {
                Console.WriteLine(read);
            }
        }
    }
}
```
the app crashes due to the OutOfMemoryException.

So, since you say that's not happening, I'm trying to understand why.

Are you using a special coreclr host and configuring how unhandled exceptions escalate?

Reply:
> the managed thread pool uses the native thread pool

Managed threadpool does not use the native thread pool in .NET Core / CoreCLR. 

Managed threadpool uses the native thread pool in .NET Native / CoreRT. Perhaps you are running into this problem with .NET Native / CoreRT only?

Reply:
I made the following toy example. Most of the time it dies. Sometimes the exception is just swallowed up.
````
using System;
using System.IO;
using System.Threading;
using System.Runtime.InteropServices;

namespace nativeoverlapped
{
    unsafe class Program
    {
        [DllImport("kernel32.dll", CharSet = CharSet.Auto)]
        internal static extern IntPtr CreateFile(string filename, uint DesiredAccess, uint ShareMode, IntPtr lpSecurityAttributes, uint creationDisposition, uint flags, IntPtr hTemplateFile);
        [DllImport("kernel32.dll")]
        internal static extern bool WriteFile(IntPtr hFile, IntPtr lpBuffer, uint numberofbytestowrite, IntPtr nullforasync, NativeOverlapped* overlapped);

        static void Main(string[] args)
        {
            AppDomain.CurrentDomain.UnhandledException += new UnhandledExceptionEventHandler(UnhandledExceptionHandler);
            var h = new Microsoft.Win32.SafeHandles.SafeFileHandle(
                CreateFile("C:\\temp\\test.dat", 3, 0, IntPtr.Zero, 2, 0x40000000, IntPtr.Zero), true);

            var b = ThreadPoolBoundHandle.BindHandle(h);
            NativeOverlapped* no = b.AllocateNativeOverlapped(delegate
            {
                throw new OutOfMemoryException();
            }, null, null);

            var bf = new byte[0x10000]; // Anything smaller and the kernel likes to make it a sync request.

            GCHandle k = new GCHandle();
            try
            {
                k = GCHandle.Alloc(bf, GCHandleType.Pinned);
                if (WriteFile(h.DangerousGetHandle(), k.AddrOfPinnedObject(), 0x10000, IntPtr.Zero, no))
                {
                    Console.WriteLine("Oops; completed synchronously.");
                }
                System.Threading.Thread.Sleep(2000);
                GC.KeepAlive(b);
                GC.KeepAlive(h);
            }
            finally
            {
                k.Free();
            }
            Console.WriteLine("End of main");
        }

        static void UnhandledExceptionHandler(object sender, UnhandledExceptionEventArgs args)
        {
            Console.WriteLine(args.ExceptionObject.ToString());
            Console.WriteLine(args.IsTerminating);
        }
    }
}
````

Reply:
> I made the following toy example. Most of the time it dies. Sometimes the exception is just swallowed up.

What .NET version are you running on?  What OS version?

I've not been able to repro the exception getting eaten.  Just to confirm, the times that the exception doesn't show up, you're not just hitting the sleep timeout? i.e. if you replace the Thread.Sleep with a Console.ReadLine, you still see the process remain up sometimes?